### ISSUE TITLE
[RISCV] BranchRelaxation calculates overly conservative size when a branch branches over compressible branch instructions

### ISSUE THREAD

**Issue by topperc** [llvm/llvm-project#56448]
The BranchFolding pass tries to determine if a branch target is too far away for the 12-bit immediate of a conditional branch. If the offset is too large it inverts the branch uses a short branch over an unconditional jump to have a larger offset.

To do this it calls getInstSize to calculate the size of all of the instructions that are branched over. If there are any compressible branches (c.bez, c.bnez), getInstSize will not calculate the compressed size for them because their offset is inherently dependent on the relaxation process. This leads to a more conservative estimate of size and can lead to more relaxation than needed.

**Comment by phyBrackets** [comment]
Hey @topperc , I'm trying to understand the issue and would like to work on it. So it currently calculates Instruction size for the compressed branch instructions , but it should not calculate right? because it depends on the relaxation?

**Comment by topperc** [comment]
A compressed branch instruction can only be generated if the branch destination is +/- 256 bytes away. The distance away is only known to the BranchRelaxation based on the size of every other instruction, the size of basic blocks, and their order, etc. When BranchRelaxation calls `RISCVInstrInfo::getInstSize()` there is no way for getInstSize to get access to that information. So I think the call to `isCompressibleInst` in `RISCVInstrInfo::getInstSize()` will assume it isn't compressible.

**Comment by phyBrackets** [comment]
Hey @topperc , I checked it with assert method if `isCompressibleInst` evaluates to true or false , it seems like it evaluates to false even for the compressed test cases , i ran all the test cases.

**Comment by phyBrackets** [comment]
Hey @topperc , could you please let me know if i'm understanding the issue right or not. 
What I understands is when for a branch instruction `getInstrSize` calculates the size for the different instructions that is in between the branch and target , if there is any compressed branch instruction in between then `getInstrSize`not able to calculate it's size which will be 2 bytes right?

**Comment by topperc** [comment]
> Hey @topperc , could you please let me know if i'm understanding the issue right or not. What I understands is when for a branch instruction `getInstrSize` calculates the size for the different instructions that is in between the branch and target , if there is any compressed branch instruction in between then `getInstrSize`not able to calculate it's size which will be 2 bytes right?

Correct. The RISCVGenCompressInstEmitter.inc will call `RISCVValidateMCOperand` and `MCOp.evaluateAsConstantImm` will fail because it isn't a constant yet.

**Comment by vsukhoml** [comment]
it seems that `.option norvc` in .S files is not handled properly on 15.0.6:
```
ld.lld: error:  init.o:(.mtrap_table+0x7a): relocation R_RISCV_BRANCH out of range: 656698 is not in [-4096, 4095]
ld.lld: warning: xxx/irq.o:(.master_handler+0x124): has non-ABS relocation R_RISCV_BRANCH against symbol '.Ltmp4'
```

**Comment by jrtc27** [comment]
Where do you see an issue with norvc there? That's a normal RVI branch.

**Comment by jrtc27** [comment]
Probably this is https://github.com/riscv-non-isa/riscv-asm-manual/pull/58 where you're relying on a surprising undocumented GNU binutils feature where the assembler will turn a single real branch instruction into its inverse and an unconditional branch as a long branch sequence.

**Comment by vsukhoml** [comment]
after few more tests it seems that issue is not in the assembler, but in ld.lld as it happens with gcc/as produced object files as well when it is used instead of ld. And in opposite, clang's output is happily processed by gnu's ld.

**Comment by jrtc27** [comment]
Are you using a linker script or relying on the linker's default layout choices (which differs between linkers)?

**Comment by vsukhoml** [comment]
yes, linker script is used. when dumping relocations I see:
```
      00000000 <__mtrap_table>:
            0: 6f 00 00 00  	j	0x0 <__mtrap_table>
                        00000000:  R_RISCV_JAL	xx_unhandled_exception
            4: 6f f0 df ff  	j	0x0 <__mtrap_table>
                        00000004:  R_RISCV_JAL	xx_unhandled_exception
            8: 6f f0 9f ff  	j	0x0 <__mtrap_table>
                        00000008:  R_RISCV_JAL	xx_unhandled_exception
            c: 6f f0 5f ff  	j	0x0 <__mtrap_table>
                        0000000c:  R_RISCV_JAL	xx_unhandled_exception
            10: 6f f0 1f ff  	j	0x0 <__mtrap_table>
                        00000010:  R_RISCV_JAL	xx_unhandled_exception
            14: 6f f0 df fe  	j	0x0 <__mtrap_table>
                        00000014:  R_RISCV_JAL	xx_unhandled_exception
            18: 6f f0 9f fe  	j	0x0 <__mtrap_table>
                        00000018:  R_RISCV_JAL	xx_unhandled_exception
            1c: 6f f0 5f fe  	j	0x0 <__mtrap_table>
                        0000001c:  R_RISCV_JAL	xx_unhandled_exception
            20: 6f f0 1f fe  	j	0x0 <__mtrap_table>
                        00000020:  R_RISCV_JAL	xx_unhandled_exception
            24: 6f f0 df fd  	j	0x0 <__mtrap_table>
                        00000024:  R_RISCV_JAL	xx_unhandled_exception
            28: 6f f0 9f fd  	j	0x0 <__mtrap_table>
                        00000028:  R_RISCV_JAL	xx_unhandled_exception
            2c: 6f f0 5f fd  	j	0x0 <__mtrap_table>
                        0000002c:  R_RISCV_JAL	xx_interrupt_from_system
            30: 6f f0 1f fd  	j	0x0 <__mtrap_table>
                        00000030:  R_RISCV_JAL	xx_unhandled_exception
            34: 6f f0 df fc  	j	0x0 <__mtrap_table>
                        00000034:  R_RISCV_JAL	xx_unhandled_exception
            38: 6f f0 9f fc  	j	0x0 <__mtrap_table>
                        00000038:  R_RISCV_JAL	xx_unhandled_exception
            3c: 6f f0 5f fc  	j	0x0 <__mtrap_table>
                        0000003c:  R_RISCV_JAL	xx_unhandled_exception
```

**Comment by jrtc27** [comment]
What is the specific relocation that is at the location indicated by LLD's error? What is the contents of your linker script? R_RISCV_JAL should stay as R_RISCV_JAL in LLD, that's not something it tries to relax AFAICT, so is unlikely to be the mysterious R_RISCV_BRANCH your error speaks of.

**Comment by vsukhoml** [comment]
so, I fixed it for now. it seems lld differs from ld in section attributes. Had add to "ax" to non-conventionally named section.
